docs/decisions/000x-internal-editor-testability-decisions.md

```diff
@@ -0,0 +1,136 @@
+
+For this purpose, we have opted for a general pattern of separating all of the behavior of components withing these editors into separate `hooks` files.
+
+These hook files contain methods that utilize both `react` and `react-redux` hooks, along with arguments passed directly into the component, in order to generate the resulting desired behaviors.
+
+From there, components are tested by mocking out the behavior of these hooks to return verifyable data in the snapshots.
+
+As part of this separation, there are a number of additional patterns that are followed 
+
+### Snapshot considerations
+#### Callbacks
+Any callback that is included in render in a component should be separated such that is is either passed in as a prop or derived from a hook, and should be mocked with a `mockName` using jest, to ensure that they are uniquely identifyable in the snapshots.
+
+Ex: 
+```javascript
+const props = {
+  onClick: jest.fn().mockName('props.onClick');
+}
+expect(shallow(<MyElement {...props} />)).toMatchSnapshot();
+```
+
+#### Imported components
+Imported compoents are mocked to return simple string components based on their existing name.  This results in shallow renders that display the components by name, with passed props, but do not attempt to render *any* logic from those components.
+
+This is a bit more complex for components with sub-components, but we have provided a test utility in `src/testUtils` called `mockNestedComponent` that will allow you to easily mock these for your snapshots as well.
+
+Ex:
+```javascript
+jest.mock('componentModule', () => {
+  const { mockNestedComponent } = jest.requireActual('./testUtils');
+  return {
+    SimpleComponents: () => 'SimpleComponent',
+    NestedComponent: mockNestedComponent('NestedComponent', {
+      NestedChild1: 'NestedComponent.NestedChild1',
+      NestedChild2: 'NestedComponent.NestedChild2',
+    }),
+});
+```
+#### Top-level mocked imports
+We have mocked out all paragon components and icons being used in the repo, as well as a number of other common methods, hooks, and components in our module's `setupTests` file, which will ensure that those components show up reasonably in snapshots.
+
+### Hook Considerations
+#### useState and mockUseState
+React's useState hook is a very powerful alternative to class components, but is also somewhat problematic to test, as it returns different values based on where it is called in a hook, as well as based on previous runs of that hook.
+
+To resolve this, we are using a custom test utility to mock a hook modules state values for easy testing.
+
+This requires a particular structure to hook modules that use the useState, for the sake of enabling the mock process (which is documented with the utility).
```

Comment:
This doesn't make sense, because we are making the code much much more complex and hard to understand. This violates the clean code principles but also is dangerous as you don't understand the state logic of the component easily and it can produce bugs. Changing implementations to make them easier testable is really not a good idea, instead, you can improve the way you test. It is absolutely possible to mock React's useState without changing the code, it is just very hard. Which is okay, because you are not supposed to mock it in general. When you need to mock it, it may be a sign that your test does not test the right thing. Unit testing means not breaking into the implementation of a component but testing it from outside, more like a user uses it. React recommends trying to test a component much like a user would - e.g. you would simulate a click on a dom element rather than overwriting the state, which is not very helpful. To make your life easier with this, I recommend react-testing-library in many cases over enzyme.
If you have trouble testing complex state logic, consider: Redux in part is designed to make this easy and make state testable. So maybe use a reducer and test that separately. Also, React has its own useReducer hooks if you don't want to interact with global state for some reason.

Reply:
That's a very good point, and I would have to figure it out a bit for myself - in previous projects I tested hooks as part of component and that's it. But there are a lot of good resources for testing custom hooks, I think even whole libraries for this purpose. A bit of research provided this very good post about it by Kent C.Dodds, who is a well-known authority in the React community, and I think his approach is very valid: https://kentcdodds.com/blog/how-to-test-custom-react-hooks

Reply:
I think testing hooks without using a React component is a bit beside the point because you are actually trying to test the component lifecycle, and that would be very far removed from reality. Any non-lifecycle related logic you can factor out of those hooks and test very easily. Personally so far I used react-testing-library (https://testing-library.com/docs/react-testing-library/intro/) for all of those cases.

Reply:
By the way the `renderHook` method from @testing-library/react Dodds uses for testing custom hooks is the same thing as react-testing-library

Reply:
a point I overlooked here in Dodd's article is "I'm not talking about the one-off custom hook you pull out just to make your component body smaller and organize your code (those should be covered by your component tests)" - I agree with that